File: Sources/LAUCaptureVideoPreviewLayer/LAUCaptureVideoPreviewLayerInternal.m
/*
 
 LAUCaptureVideoPreviewLayerInternal.m
 LAUCaptureVideoPreviewLayer
 
 Copyright (c) 2016 Luis Laugga.
 Some rights reserved, all wrongs deserved.
 
 Permission is hereby granted, free of charge, to any person obtaining a copy of
 this software and associated documentation files (the "Software"), to deal in
 the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
*/

#import "LAUCaptureVideoPreviewLayerInternal.h"
#import "LAUCaptureVideoPreviewLayerLogging.h"

#define kVideoDataOutputSampleBuffersSize 2

@interface LAUCaptureVideoPreviewLayerInternal () <AVCaptureVideoDataOutputSampleBufferDelegate>
{
    // Session
    AVCaptureSession * _session;
    
    // AVCaptureVideoDataOutput and delegate queue
    AVCaptureVideoDataOutput * _videoDataOutput;
    dispatch_queue_t _videoDataOutputSampleBufferDelegateQueue;
    
    // Circular array used to keep the sample buffers
    NSUInteger _videoDataOutputSampleBuffersHeadIndex;
    NSUInteger _videoDataOutputSampleBuffersTailIndex;
    CMSampleBufferRef _videoDataOutputSampleBuffers[kVideoDataOutputSampleBuffersSize];
    
    // Hijacked AVCaptureVideoDataOutput (check
    dispatch_queue_t _hijackedVideoDataOutputSampleBufferDelegateQueue;
    id <AVCaptureVideoDataOutputSampleBufferDelegate> _hijackedVideoDataOutputSampleBufferDelegate;
}
@end

@implementation LAUCaptureVideoPreviewLayerInternal

#pragma mark -
#pragma mark AVCaptureSession

- (instancetype)init
{
    self = [super init];
    if (self)
    {
        _videoDataOutputSampleBuffersHeadIndex = _videoDataOutputSampleBuffersTailIndex = 0;
        memset(_videoDataOutputSampleBuffers, 0, kVideoDataOutputSampleBuffersSize);
    }
    return self;
}

- (void)setSession:(AVCaptureSession *)session
{
    if (_session)
    {
        // Stop observing the old session
        [[NSNotificationCenter defaultCenter] removeObserver:self name:nil object:_session];
    }
    
    _session = session;
    
    // Set the session's AVCaptureVideoDataOutput
    if ([_session canAddOutput:[self videoDataOutput]])
    {
        Log(@"LAUCaptureVideoPreviewLayerInternal: Added AVCaptureVideoDataOutput to AVCaptureSession");
        [_session addOutput:[self videoDataOutput]];
    }
    else
    {
        // TODO improve this. After setSession the hijackedSession with prevent canAddOutput...
        
        Log(@"LAUCaptureVideoPreviewLayerInternal: Can NOT add AVCaptureVideoDataOutput to AVCaptureSession");
        [self hijackSessionVideoDataOutput];
    }
    
    // Observe for specific notifications related with the session
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sessionDidPostNotification:) name:nil object:_session];
}

- (void)hijackSessionVideoDataOutput
{
    Log(@"LAUCaptureVideoPreviewLayerInternal: Hijacking current AVCaptureVideoDataOutput of AVCaptureSession");
    
    if (_session)
    {
        AVCaptureVideoDataOutput * currentVideoDataOutput = nil;
        
        // Look for an instance of AVCaptureVideoDataOutput in session's outputs
        for (AVCaptureOutput * currentOutput in _session.outputs)
        {
            if ([currentOutput isKindOfClass:[AVCaptureVideoDataOutput class]])
            {
                currentVideoDataOutput = (AVCaptureVideoDataOutput *)currentOutput;
                break;
            }
        }
        
        if (currentVideoDataOutput)
        {
            // Copy a reference of current delegate and queue
            _hijackedVideoDataOutputSampleBufferDelegate = currentVideoDataOutput.sampleBufferDelegate;
            _hijackedVideoDataOutputSampleBufferDelegateQueue = currentVideoDataOutput.sampleBufferCallbackQueue;
            
            // we want BGRA, both CoreGraphics and Metal work well with 'BGRA'
            NSDictionary * videoSettings = @{(id)kCVPixelBufferPixelFormatTypeKey:@(kCMPixelFormat_32BGRA)};
            [currentVideoDataOutput setVideoSettings:videoSettings];
            [currentVideoDataOutput setAlwaysDiscardsLateVideoFrames:YES]; // discard if the data output queue is blocked
            
            // Set the video data output delegate
            [currentVideoDataOutput setSampleBufferDelegate:self queue:[self videoDataOutputSampleBufferDelegateQueue]];
            
            // Enable the video data output
            [[currentVideoDataOutput connectionWithMediaType:AVMediaTypeVideo] setEnabled:YES];
        }
        
    }
}

#pragma mark -
#pragma mark AVCaptureSession Notifications

- (void)sessionDidPostNotification:(NSNotification *)notification
{
    if ([notification.name isEqualToString:AVCaptureSessionWasInterruptedNotification] ||
        [notification.name isEqualToString:AVCaptureSessionRuntimeErrorNotification] ||
        [notification.name isEqualToString:AVCaptureSessionDidStopRunningNotification])
    {
        if ([_delegate respondsToSelector:@selector(captureVideoPreviewLayerInternal:sessionDidStopRunning:)]) {
            [_delegate captureVideoPreviewLayerInternal:self sessionDidStopRunning:_session];
        }
    }
    else
    {
        if ([_delegate respondsToSelector:@selector(captureVideoPreviewLayerInternal:sessionDidStartRunning:)]) {
            [_delegate captureVideoPreviewLayerInternal:self sessionDidStartRunning:_session];
        }
    }
}

#pragma mark -
#pragma mark AVCaptureVideoDataOutput

- (dispatch_queue_t)videoDataOutputSampleBufferDelegateQueue
{
    if (!_videoDataOutputSampleBufferDelegateQueue)
    {
        // Create a serial dispatch queue used for the sample buffer delegate.
        // In a multi-threaded producer consumer system it's generally a good idea to make sure that producers
        // do not get starved of CPU time by their consumers. In this app we start with VideoDataOutput frames
        // on a high priority queue, and downstream consumers use default priority queues.
        dispatch_queue_t captureOutputSampleBufferDelegateQueue = dispatch_queue_create( "co.coletiv.lightmate.LAUCaptureVideoPreviewLayerInternal.videoDataOutputSampleBufferDelegateQueue", DISPATCH_QUEUE_SERIAL);
        dispatch_set_target_queue(captureOutputSampleBufferDelegateQueue, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0));
        
        _videoDataOutputSampleBufferDelegateQueue = captureOutputSampleBufferDelegateQueue;
    }
    
    return _videoDataOutputSampleBufferDelegateQueue;
}

- (AVCaptureVideoDataOutput *)videoDataOutput
{
    if (!_videoDataOutput)
    {
        AVCaptureVideoDataOutput * videoDataOutput = [AVCaptureVideoDataOutput new];
        [self configureVideoDataOutput:videoDataOutput];
        
        _videoDataOutput = videoDataOutput;
    }
    
    return _videoDataOutput;
}

- (void)configureVideoDataOutput:(AVCaptureVideoDataOutput *)videoDataOutput
{
    // we want BGRA, both CoreGraphics and Metal work well with 'BGRA'
    NSDictionary * videoSettings = @{(id)kCVPixelBufferPixelFormatTypeKey:@(kCMPixelFormat_32BGRA)};
    [videoDataOutput setVideoSettings:videoSettings];
    [videoDataOutput setAlwaysDiscardsLateVideoFrames:YES]; // discard if the data output queue is blocked
    
    // Set the video data output delegate
    [videoDataOutput setSampleBufferDelegate:self queue:[self videoDataOutputSampleBufferDelegateQueue]];
    
    // Enable the video data output
    [[videoDataOutput connectionWithMediaType:AVMediaTypeVideo] setEnabled:YES];
}

#pragma mark -
#pragma mark AVCaptureVideoDataOutputSampleBufferDelegate

- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection
{
    //PrettyLog;
    
    // Add the sample buffer to the _videoDataOutputSampleBuffers circular array
    [self addSampleBuffer:sampleBuffer];
    
    // Was the AVCaptureVideoDataOutput's hijacked?
    if (_hijackedVideoDataOutputSampleBufferDelegate && _hijackedVideoDataOutputSampleBufferDelegateQueue)
    {
        dispatch_async(_hijackedVideoDataOutputSampleBufferDelegateQueue, ^{
            
            // Forward the delegate method to the AVCaptureVideoDataOutput's hijacked sampleBufferDelegate
            if ([_hijackedVideoDataOutputSampleBufferDelegate respondsToSelector:@selector(captureOutput:didOutputSampleBuffer:fromConnection:)])
            {
                [_hijackedVideoDataOutputSampleBufferDelegate captureOutput:captureOutput didOutputSampleBuffer:sampleBuffer fromConnection:connection];
            }
        });
    }
}

#pragma mark -
#pragma mark Ping-pong buffering for videoDataOutputSampleBuffers

- (CMSampleBufferRef)sampleBuffer
{
    @synchronized (self)
    {
        //PrettyLog;
        
        CMSampleBufferRef videoDataOutputSampleBuffer = NULL;
        
        // Check if the circular array is not empty
        if (_videoDataOutputSampleBuffersHeadIndex != _videoDataOutputSampleBuffersTailIndex)
        {
            // Remove an existing sample buffer from the head
            CMSampleBufferRef oldVideoDataOutputSampleBuffer = _videoDataOutputSampleBuffers[_videoDataOutputSampleBuffersHeadIndex];
            
            // Release the buffer if it exists
            if(oldVideoDataOutputSampleBuffer != NULL)
            {
                videoDataOutputSampleBuffer = oldVideoDataOutputSampleBuffer;
            }
            
            //Log(@"\nsampleBuffer head %d tail %d info %@\n\n\n", _videoDataOutputSampleBuffersHeadIndex, _videoDataOutputSampleBuffersTailIndex, videoDataOutputSampleBuffer);
            
            // Move head +1
            _videoDataOutputSampleBuffersHeadIndex = (_videoDataOutputSampleBuffersHeadIndex+1)%kVideoDataOutputSampleBuffersSize;
        }
        
        return videoDataOutputSampleBuffer;
    }
}

- (void)addSampleBuffer:(CMSampleBufferRef)newVideoDataOutputSampleBuffer
{
    @synchronized (self)
    {
        //PrettyLog;
        
        if (newVideoDataOutputSampleBuffer != NULL)
        {
            // Add the new sample buffer to the tail
            CMSampleBufferRef oldVideoDataOutputSampleBuffer = _videoDataOutputSampleBuffers[_videoDataOutputSampleBuffersTailIndex];
            
            // Release an old buffer if it exists
            if(oldVideoDataOutputSampleBuffer != NULL)
            {
                // Release old
                CFRelease(oldVideoDataOutputSampleBuffer);
            }
            
            // Retain new
            _videoDataOutputSampleBuffers[_videoDataOutputSampleBuffersTailIndex] = newVideoDataOutputSampleBuffer;
            CFRetain(newVideoDataOutputSampleBuffer);
            
            // Move tail +1
            _videoDataOutputSampleBuffersTailIndex = (_videoDataOutputSampleBuffersTailIndex+1)%kVideoDataOutputSampleBuffersSize;
            
            // As a circular array we make sure old unused sample buffers are discarded by moving head forwards +1
            if (_videoDataOutputSampleBuffersTailIndex == _videoDataOutputSampleBuffersHeadIndex)
            {
                // Move head +1
                _videoDataOutputSampleBuffersHeadIndex = (_videoDataOutputSampleBuffersTailIndex+1)%kVideoDataOutputSampleBuffersSize;
            }
            
            //Log(@"\naddSampleBuffer head %d tail %d info %@\n\n\n", _videoDataOutputSampleBuffersHeadIndex, _videoDataOutputSampleBuffersTailIndex, newVideoDataOutputSampleBuffer);
        }
    }
}

- (void)flushSampleBuffer
{
    @synchronized (self)
    {
        while(_videoDataOutputSampleBuffersHeadIndex != _videoDataOutputSampleBuffersTailIndex)
        {
            // Remove an existing sample buffer from the head
            CMSampleBufferRef videoDataOutputSampleBuffer = _videoDataOutputSampleBuffers[_videoDataOutputSampleBuffersHeadIndex];
            
            // Release the buffer if it exists
            if(videoDataOutputSampleBuffer != NULL)
            {
                CFRelease(videoDataOutputSampleBuffer);
                _videoDataOutputSampleBuffers[_videoDataOutputSampleBuffersHeadIndex] = NULL;
            }
            
            // Move head +1
            _videoDataOutputSampleBuffersHeadIndex = (_videoDataOutputSampleBuffersHeadIndex+1)%kVideoDataOutputSampleBuffersSize;
        }
    }
}

@end

File: Sources/LAUCaptureVideoPreviewLayer/LAUCaptureVideoPreviewLayerInternal.swift
/*

 LAUCaptureVideoPreviewLayerInternal.swift
 LAUCaptureVideoPreviewLayer

 Copyright (c) 2016 Luis Laugga.
 Some rights reserved, all wrongs deserved.

 Permission is hereby granted, free of charge, to any person obtaining a copy of
 this software and associated documentation files (the "Software"), to deal in
 the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

*/

import AVFoundation

/// Informs the delegate object about state changes in the capture pipeline.
public protocol LAUCaptureVideoPreviewLayerInternalDelegate: AnyObject {

    func captureVideoPreviewLayerInternal(_ internalPipeline: LAUCaptureVideoPreviewLayerInternal,
                                          sessionDidStopRunning session: AVCaptureSession?)

    func captureVideoPreviewLayerInternal(_ internalPipeline: LAUCaptureVideoPreviewLayerInternal,
                                          sessionDidStartRunning session: AVCaptureSession?)
}

/// The capture side of the preview layer.
///
/// It adds an `AVCaptureVideoDataOutput` to the session, or hijacks the one already
/// there, and keeps the most recent sample buffers in a small circular array for the
/// renderer to pick up.
///
/// It is open so that it can be replaced by a stand-in that feeds the preview layer
/// without a capture device, which is what the tests and the example do.
open class LAUCaptureVideoPreviewLayerInternal: NSObject {

    private static let sampleBuffersSize = 2

    /// The AVCaptureSession instance being previewed by the receiver.
    open var session: AVCaptureSession? {
        didSet {
            sessionDidChange(from: oldValue)
        }
    }

    /// The sample buffer being currently displayed.
    open var sampleBuffer: CMSampleBuffer? {
        lock.lock()
        defer { lock.unlock() }

        // Check if the circular array is not empty
        guard headIndex != tailIndex else {
            return nil
        }

        // Remove an existing sample buffer from the head
        let sampleBuffer = sampleBuffers[headIndex]

        // Move head +1
        headIndex = (headIndex + 1) % Self.sampleBuffersSize

        return sampleBuffer
    }

    /// YES if the session is running. NO for all other possible states:
    /// RuntimeError, Interrupted, Stopped.
    open var sessionIsRunning: Bool {
        session?.isRunning ?? false
    }

    open weak var delegate: LAUCaptureVideoPreviewLayerInternalDelegate?

    // AVCaptureVideoDataOutput and delegate queue
    private var videoDataOutput: AVCaptureVideoDataOutput?
    private lazy var videoDataOutputSampleBufferDelegateQueue: DispatchQueue = {
        // In a multi-threaded producer consumer system it's generally a good idea to
        // make sure that producers do not get starved of CPU time by their consumers.
        // We start with VideoDataOutput frames on a high priority queue, and
        // downstream consumers use default priority queues.
        DispatchQueue(label: "com.laugga.LAUCaptureVideoPreviewLayerInternal.videoDataOutputSampleBufferDelegateQueue",
                      qos: .userInitiated)
    }()

    // Circular array used to keep the sample buffers
    private let lock = NSLock()
    private var headIndex = 0
    private var tailIndex = 0
    private var sampleBuffers = [CMSampleBuffer?](repeating: nil, count: sampleBuffersSize)

    // Hijacked AVCaptureVideoDataOutput
    private weak var hijackedSampleBufferDelegate: AVCaptureVideoDataOutputSampleBufferDelegate?
    private var hijackedSampleBufferDelegateQueue: DispatchQueue?

    public override init() {
        super.init()
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    // MARK: - AVCaptureSession

    private func sessionDidChange(from oldSession: AVCaptureSession?) {

        if let oldSession {
            // Stop observing the old session
            NotificationCenter.default.removeObserver(self, name: nil, object: oldSession)
        }

        guard let session else {
            return
        }

        // Set the session's AVCaptureVideoDataOutput
        let videoDataOutput = makeVideoDataOutput()

        if session.canAddOutput(videoDataOutput) {
            log.debug("Added AVCaptureVideoDataOutput to AVCaptureSession")
            session.addOutput(videoDataOutput)
            configure(videoDataOutput)
            self.videoDataOutput = videoDataOutput
        } else {
            // TODO improve this. After setSession the hijackedSession will prevent canAddOutput...
            log.debug("Can NOT add AVCaptureVideoDataOutput to AVCaptureSession")
            hijackSessionVideoDataOutput()
        }

        // Observe for specific notifications related with the session
        for name in [AVCaptureSession.wasInterruptedNotification,
                     AVCaptureSession.runtimeErrorNotification,
                     AVCaptureSession.didStopRunningNotification,
                     AVCaptureSession.didStartRunningNotification,
                     AVCaptureSession.interruptionEndedNotification] {
            NotificationCenter.default.addObserver(self,
                                                   selector: #selector(sessionDidPostNotification(_:)),
                                                   name: name,
                                                   object: session)
        }
    }

    private func hijackSessionVideoDataOutput() {

        log.debug("Hijacking current AVCaptureVideoDataOutput of AVCaptureSession")

        // Look for an instance of AVCaptureVideoDataOutput in session's outputs
        guard let currentVideoDataOutput = session?.outputs.compactMap({ $0 as? AVCaptureVideoDataOutput }).first else {
            return
        }

        // Copy a reference of current delegate and queue
        hijackedSampleBufferDelegate = currentVideoDataOutput.sampleBufferDelegate
        hijackedSampleBufferDelegateQueue = currentVideoDataOutput.sampleBufferCallbackQueue

        configure(currentVideoDataOutput)

        videoDataOutput = currentVideoDataOutput
    }

    // MARK: - AVCaptureSession notifications

    @objc private func sessionDidPostNotification(_ notification: Notification) {

        switch notification.name {
        case AVCaptureSession.wasInterruptedNotification,
             AVCaptureSession.runtimeErrorNotification,
             AVCaptureSession.didStopRunningNotification:
            delegate?.captureVideoPreviewLayerInternal(self, sessionDidStopRunning: session)
        default:
            delegate?.captureVideoPreviewLayerInternal(self, sessionDidStartRunning: session)
        }
    }

    // MARK: - AVCaptureVideoDataOutput

    private func makeVideoDataOutput() -> AVCaptureVideoDataOutput {
        videoDataOutput ?? AVCaptureVideoDataOutput()
    }

    private func configure(_ videoDataOutput: AVCaptureVideoDataOutput) {

        // We want BGRA, both CoreGraphics and Metal work well with 'BGRA'
        videoDataOutput.videoSettings = [kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA]

        // Discard if the data output queue is blocked
        videoDataOutput.alwaysDiscardsLateVideoFrames = true

        // Set the video data output delegate
        videoDataOutput.setSampleBufferDelegate(self, queue: videoDataOutputSampleBufferDelegateQueue)

        // Enable the video data output
        videoDataOutput.connection(with: .video)?.isEnabled = true
    }

    // MARK: - Ping-pong buffering for the video data output sample buffers

    private func add(_ newSampleBuffer: CMSampleBuffer) {
        lock.lock()
        defer { lock.unlock() }

        // Add the new sample buffer to the tail
        sampleBuffers[tailIndex] = newSampleBuffer

        // Move tail +1
        tailIndex = (tailIndex + 1) % Self.sampleBuffersSize

        // As a circular array we make sure old unused sample buffers are discarded by
        // moving head forwards +1
        if tailIndex == headIndex {
            headIndex = (tailIndex + 1) % Self.sampleBuffersSize
        }
    }

    open func flushSampleBuffer() {
        lock.lock()
        defer { lock.unlock() }

        sampleBuffers = [CMSampleBuffer?](repeating: nil, count: Self.sampleBuffersSize)
        headIndex = 0
        tailIndex = 0
    }
}

// MARK: - AVCaptureVideoDataOutputSampleBufferDelegate

extension LAUCaptureVideoPreviewLayerInternal: AVCaptureVideoDataOutputSampleBufferDelegate {

    public func captureOutput(_ output: AVCaptureOutput,
                              didOutput sampleBuffer: CMSampleBuffer,
                              from connection: AVCaptureConnection) {

        // Add the sample buffer to the circular array
        add(sampleBuffer)

        // Was the AVCaptureVideoDataOutput hijacked?
        guard let hijackedSampleBufferDelegate,
              let hijackedSampleBufferDelegateQueue else {
            return
        }

        hijackedSampleBufferDelegateQueue.async {
            // Forward the delegate method to the hijacked sample buffer delegate
            hijackedSampleBufferDelegate.captureOutput?(output, didOutput: sampleBuffer, from: connection)
        }
    }
}
